R/utils.R

```diff
@@ -0,0 +1,146 @@
+#' that are important to us for tag writing:
+#'
+#' 1. Text is automatically converted to UTF-8.
+#'
+#' 2. Ability to bookmark (save) a write position, and jump back
+#'    to it (restore). Only one bookmark at a time is allowed.
+#'    Restoring a bookmark essentially truncates content that
+#'    was written after the bookmark was set.
+#'
+#' 3. Ability to suppress writes (setSuppress); while suppression
+#'    is set to true, all write operations no-op.
+#'
+#' @noRd
+#' @importFrom R6 R6Class
+TextWriter <- R6Class("TextWriter",
+  private = list(
+    con = "ANY",
+    marked = numeric(1),
+    suppressing = logical(1)
+  ),
+  public = list(
+    initialize = function() {
+      # We use a file() here instead of textConnection() or paste/c to
+      # avoid the overhead of copying, which is huge for moderately
+      # large numbers of calls to .$write(). Generally when you want
+      # to incrementally build up a long string out of immutable ones,
+      # you want to use a mutable/growable string buffer of some kind;
+      # since R doesn't have something like that (that I know of),
+      # file() is the next best thing.
+      private$con <- file("", "w+b", encoding = "UTF-8")
+      private$marked <- 0
+      private$suppressing <- FALSE
+    },
+    close = function() {
+      close(private$con)
+    },
+    # Write content; may or may not contain incidental whitespace,
+    # but whitespace that needs to be suppressible (like newlines
+    # and indents that the user did not explicitly request) should
+    # use writeWS() instead.
+    #
+    # Calling write() causes suppressWhitespace() to be cancelled.
+    #
+    # @param text Single element character vector
+    write = function(text) {
+      if (!private$suppressing) {
+        # The text that is written to this TextWriter will be converted to
+        # UTF-8 using enc2utf8. The rendered output will always be UTF-8
+        # encoded.
+
+        raw <- charToRaw(enc2utf8(text))
+        # This is actually writing UTF-8 bytes, not chars
+        writeBin(raw, private$con)
+      }
+    },
+    # Return the contents of the TextWriter, as a single element
+    # character vector, from the beginning to the current writing
+    # position (normally this is the end of the connection, unless
+    # restorePosition() was called).
+    readAll = function() {
+      wpos <- seek(private$con, where = NA, rw = "write")
+      seek(private$con, where = 0, origin = "start", rw = "read")
+      s <- readChar(private$con, wpos, useBytes = TRUE)
+      Encoding(s) <- "UTF-8"
+      s
+    },
+    # Mark the current writing position
+    savePosition = function() {
+      # Save the current write pos
+      private$marked <- seek(private$con, where = NA, rw = "write")
+    },
+    # Jump to the most recently marked position
+    restorePosition = function() {
+      if (!is.na(private$marked)) {
+        seek(private$con, private$marked, origin = "start", rw = "write")
+      }
```

Comment:
else error? warn?

Reply:
Maybe? I'm not even sure why there's a check there, it doesn't look to me like private$marked could be anything but a non-NA number? Maybe the original default for marked was NA, but then I changed it to 0.